### ISSUE TITLE
[clsim] I3CLSimLightSourceToStepConverterFlasher hangs (Trac #2426)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2418]
Steps to reproduce:

```text
clsim/resources/scripts/flasher/generateTestFlashes.py
clsim/resources/scripts/flasher/applyCLSim.py -i test_flashes.i3
```

and observe the second process not exiting, and also not using any CPU time.

This happens `I3CLSimLightSourceToStepConverterFlasher` can only run in the main thread, as it attempts to call `I3CLSimRandomValueIceCubeFlasherTimeProfile.SampleFromDistribution()` and waits forever trying to acquire the GIL that the main thread never releases. This clashes with its use within I3CLSimLightSourceToStepConverterAsync.

There are two solutions: either drop the GIL in `I3Tray::Execute()` (deja vu, anyone?) or port I3CLSimRandomValueIceCubeFlasherTimeProfile to C++. The latter is probably easier in the short term.

Migrated from https://code.icecube.wisc.edu/ticket/2426
```json
{
    "status": "closed", 
    "changetime": "2020-06-24T12:32:17", 
    "description": "Steps to reproduce:\n\n{{{\nclsim/resources/scripts/flasher/generateTestFlashes.py\nclsim/resources/scripts/flasher/applyCLSim.py -i test_flashes.i3\n}}}\n\nand observe the second process not exiting, and also not using any CPU time.\n\nThis happens `I3CLSimLightSourceToStepConverterFlasher` can only run in the main thread, as it attempts to call `I3CLSimRandomValueIceCubeFlasherTimeProfile.SampleFromDistribution()` and waits forever trying to acquire the GIL that the main thread never releases. This clashes with its use within I3CLSimLightSourceToStepConverterAsync.\n\nThere are two solutions: either drop the GIL in `I3Tray::Execute()` (deja vu, anyone?) or port I3CLSimRandomValueIceCubeFlasherTimeProfile to C++. The latter is probably easier in the short term.", 
    "reporter": "jvansanten", 
    "cc": "fiedl", 
    "resolution": "fixed", 
    "_ts": "1593001937450890", 
    "component": "combo simulation", 
    "summary": "[clsim] I3CLSimLightSourceToStepConverterFlasher hangs", 
    "priority": "major", 
    "keywords": "", 
    "time": "2020-04-23T19:42:44", 
    "milestone": "Autumnal Equinox 2020", 
    "owner": "jvansanten", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20200503T08:36:50: 
* jvansanten changed resolution from "" to "fixed"
* jvansanten changed status from "new" to "closed"
* jvansanten commented:

> In changeset 179989:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="179989"
> Release the GIL in I3CLSimClientModule
> 
> Fixes #2418. Note that while all function calls from C++ back into
> Python in clsim are protected with GIL guards, objects allocated in
> Python whose __dict__ contains other Python objects will segfault if
> freed in a thread that does not hold the GIL, as noted in the still-
> unresolved https://github.com/boostorg/python/pull/11. There's also
> no good way around this, short of keeping a (potentially unbounded)
> queue of PyObjects to be decref'd when the GIL is next acquired.
> 
> Reacquiring the GIL before returning to I3Module::Process_()
> should however prevent most of these cases, as the GIL will be held
> when any dropped frames go out of scope.
> ```


* jvansanten changed owner from "" to "jvansanten"

**Comment by icecube-trac** [comment]
Trac update at 20200624T12:32:17: 
* olivas changed milestone from "Summer Solstice 2020" to "Autumnal Equinox 2020"
* olivas commented:

> Milestone Summer Solstice 2020 deleted
